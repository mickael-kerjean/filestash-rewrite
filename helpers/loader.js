import { get as getRelease } from "../pages/adminpage/model_release.js";

let version = null;

export async function loadScript(url) {
    const $script = document.createElement("script");
    $script.setAttribute("src", url);
    document.head.appendChild($script);
    return new Promise((done) => {
        $script.onload = done;
        $script.onerror = done;
    });
}

export async function CSS(baseURL, ...arrayOfFilenames) {
    const sheets = await Promise.all(arrayOfFilenames.map((filename) => loadSingleCSS(baseURL, filename)));
    return sheets.join("\n\n");
}

async function loadSingleCSS(baseURL, filename) {
    const res = await fetch(baseURL.replace(/(.*)\/[^\/]+$/, "$1/") + `${filename}?version=${version}`, {
        cache: "force-cache",
    });
    if (res.status !== 200) return `/* ERROR: ${res.status} */`;
    else if (!res.headers.get("Content-Type")?.startsWith("text/css")) return `/* ERROR: wrong type, got "${res.headers?.get("Content-Type")}"*/`;
    return await res.text();
}

export async function init() {
    const info = await getRelease().toPromise();
    version = info.version;
}
