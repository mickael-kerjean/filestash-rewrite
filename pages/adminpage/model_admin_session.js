import rxjs from "../../lib/rx.js";
import ajax from "../../lib/ajax.js";

const sessionSubject$ = new rxjs.Subject(1);

const adminSession$ = rxjs.merge(
    sessionSubject$,
    rxjs.interval(30000).pipe(
        rxjs.startWith(null),
        rxjs.mergeMap(() => ajax({ url: "/admin/api/session", responseType: "json" })),
        rxjs.map(({ responseJSON }) => responseJSON.result),
        rxjs.distinctUntilChanged(),
    )
).pipe(rxjs.shareReplay(1));

export function isAdmin$() {
    return adminSession$;
}

export function authenticate$(body) {
    return ajax({
        url: "/admin/api/session",
        method: "POST",
        body,
        responseType: "json"
    }).pipe(
        rxjs.mapTo(true),
        rxjs.tap((ok) => ok && sessionSubject$.next(ok)),
    );
}
