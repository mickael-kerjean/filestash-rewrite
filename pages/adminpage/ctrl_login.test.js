import rxjs from "../../lib/rx.js";
import ctrl from "./ctrl_login.js";
import { CSS } from "../../helpers/loader.js";

import { authenticate$ } from "./model_admin_session.js";

jest.mock("./model_admin_session.js");
jest.mock("../../helpers/loader.js");

describe("admin::ctrl_login", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        CSS.mockImplementation(() => Promise.resolve(""));
    });
    it("renders a form", async () => {
        // given
        const render = jest.fn();

        // when
        await ctrl(render);

        // then
        const $page = render.mock.calls[0][0];
        expect($page).toBeTruthy();
        expect($page.querySelector("form")).toBeTruthy();
        expect($page.querySelector(`form input[type="password"]`)).toBeTruthy();
    });

    it("snapshot", async () => {
        // given
        const render = jest.fn();

        // when
        await ctrl(render);

        // then
        expect(render.mock.calls[0][0]).toMatchSnapshot();
    });

    it("calls authentication logic on submit", async () => {
        // given
        let formData = null;
        let $page = null;
        authenticate$.mockImplementation((body) => rxjs.of(body).pipe(
            rxjs.tap((_formData) => formData = _formData)
        ));
        await ctrl(($node) => $page = $node);
        const $password = $page.querySelector(`form input[type="password"]`)
        const $form = $page.querySelector("form");

        // when
        $password.value = "__password__";
        $form.dispatchEvent(new window.Event("submit"));

        // then
        expect(authenticate$).toBeCalledTimes(1);
        expect(formData).toStrictEqual({password: "__password__"});
    });
});
