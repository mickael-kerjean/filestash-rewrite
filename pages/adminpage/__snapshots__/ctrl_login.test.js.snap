// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin::ctrl_login snapshot 1`] = `
<div
  class="component_container component_page_adminlogin"
  style="margin-top: 256px;"
>
  
            
  <style />
  
            
  <form>
    
                
    <div
      class="input_group"
    >
      
                    
      <input
        autocomplete=""
        class="component_input"
        name="password"
        placeholder="Password"
        type="password"
      />
      
                    
      <button
        class="transparent"
      >
        
                        
        <component-icon
          name="arrow_right"
        />
        
                    
      </button>
      
                
    </div>
    
            
  </form>
  
        
</div>
`;
